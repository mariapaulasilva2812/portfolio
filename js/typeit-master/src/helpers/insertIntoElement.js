import isLastAtEveryLevel from "./isLastAtEveryLevel";
import createNode from "./createNode";
import nodeCollectionToArray from "./nodeCollectionToArray";
import isInput from "./isInput";

/**
 * Inserts a set of content into the element. Intended for SINGLE characters.
 *
 * @param {object} element
 * @param {string | object} content
 */
export default (element, contentArg) => {
  // Assume it's a string, and maybe overwrite later.
  let content = contentArg;

  if (isInput(element)) {
    element.value = `${element.value}${content}`;
    return;
  }

  // Find any existing cursor in the element to make sure we type BEFORE it.
  let cursorNode = nodeCollectionToArray(element.childNodes).filter(n => {
    return n.classList && n.classList.contains("ti-cursor");
  });

  cursorNode = cursorNode.length ? cursorNode[0] : null;

  // We're inserting a character within an element!
  // Make sure this isn't an HTML node that's being inserted.
  if (typeof contentArg === "object" && !(contentArg instanceof HTMLElement)) {
    let ancestorTree = contentArg.ancestorTree.slice(0);
    let parentSelectors = ancestorTree.reverse().join(" ");
    let existingNodes = nodeCollectionToArray(
      element.querySelectorAll(`${parentSelectors}`)
    );
    let lastExistingNode = existingNodes[existingNodes.length - 1];

    // Only type into an existing element if there is one
    // and it's the last one in the entire container.
    if (lastExistingNode && isLastAtEveryLevel(lastExistingNode, cursorNode)) {
      element = lastExistingNode;
      content = contentArg.content;

      // We need to create an element!
    } else {
      // Overwrite the content with this newly created element.
      content = createNode(
        contentArg.ancestorTree[0],
        contentArg.attributes,
        contentArg.content
      );

      // We know this new element is supposed to be nested, so we need to print it inside the
      // the LAST parent node that was created inside the container.
      if (contentArg.ancestorTree.length > 1) {
        // Get all the parent nodes in the container.
        let parentNodes = nodeCollectionToArray(
          element.querySelectorAll(contentArg.ancestorTree[1])
        );

        // We assume it exists.
        element = parentNodes[parentNodes.length - 1];
      }
    }
  }

  // Might be either an HTMLElement or Node.
  content =
    typeof content === "object" ? content : document.createTextNode(content);

  // If a cursor node exists, make sure we print BEFORE that, but only if the target
  // element is the top-level one. Otherwise, stick it to the end of the element.
  element.insertBefore(
    content,
    cursorNode && element.hasAttribute("data-typeit-id") ? cursorNode : null
  );
};
