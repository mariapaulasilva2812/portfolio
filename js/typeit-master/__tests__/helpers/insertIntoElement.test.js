import insertIntoElement from "../../src/helpers/insertIntoElement";

let someElement;

beforeEach(() => {
  someElement = document.createElement("SPAN");
});

test("Should insert a simple character correctly.", () => {
  insertIntoElement(someElement, "x");
  expect(someElement.innerHTML).toBe(`x`);
});

test("Should insert a character object.", () => {
  let characterObject = {
    ancestorTree: ["SPAN"],
    attributes: [],
    content: "y",
    isFirstChar: true
  };
  insertIntoElement(someElement, characterObject);

  expect(someElement.innerHTML).toBe(`<span>y</span>`);
});

test("Should insert a nested character object.", () => {
  let characterObject = {
    ancestorTree: ["SPAN", "EM"],
    attributes: [],
    content: "y",
    isFirstChar: false
  };
  someElement.innerHTML = `<em></em>`;
  insertIntoElement(someElement, characterObject);
  expect(someElement.innerHTML).toMatchSnapshot();
});

test("Should insert content into input.", () => {
  setHTML(`
    <div>
      <input id="inputElement" type="text" />
    </div>
  `);

  someElement = document.getElementById("inputElement");

  insertIntoElement(someElement, "some value");

  expect(someElement.value).toBe("some value");
});

test("Should insert raw HTML content into input.", () => {
  setHTML`
    <div>
      <input id="inputElement" type="text" />
    </div>
  `;

  someElement = document.getElementById("inputElement");

  insertIntoElement(someElement, "<span>sup</span>");

  expect(someElement.value).toBe("<span>sup</span>");
});

test("Should insert before cursor when element is top-level.", () => {
  setHTML`
    <span data-typeit-id="9">
      <i class="ti-cursor">|</i>
    </span>
  `;

  someElement = document.querySelector("span");

  insertIntoElement(someElement, "Hello");

  expect(document.body.innerHTML).toMatchSnapshot();
});

test("Should not insert before cursor when target element is not top-level.", () => {
  setHTML`
    <span>
      <i class="ti-cursor">|</i>
    </span>
  `;

  someElement = document.querySelector("span");

  insertIntoElement(someElement, "Hello");

  expect(document.body.innerHTML).toMatchSnapshot();
});
