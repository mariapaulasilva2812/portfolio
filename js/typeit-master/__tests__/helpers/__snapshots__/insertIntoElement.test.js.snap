// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Should insert a nested character object. 1`] = `"<em><span>y</span></em>"`;

exports[`Should insert before cursor when element is top-level. 1`] = `"    <span data-typeit-id=\\"9\\">Hello<i class=\\"ti-cursor\\">|</i></span>  "`;

exports[`Should not insert before cursor when target element is not top-level. 1`] = `"    <span><i class=\\"ti-cursor\\">|</i>Hello</span>  "`;
